BridgeTower blog post (#1118)
huggingface/blog: bridgetower.md

[Habana Gaudi2](https://habana.ai/products/gaudi2/) is the second-generation AI hardware accelerator designed by Habana Labs. A single server contains 8 accelerator devices called HPUs with 96GB of memory each. Check out [this previous blog post](https://huggingface.co/blog/habana-gaudi-2-bloom#habana-gaudi2) for a more in-depth presentation and a guide showing how to access it through the [Intel Developer Cloud](https://www.intel.com/content/www/us/en/secure/developer/devcloud/cloud-launchpad.html). Furthermore, unlike many AI accelerators on the market, using advanced features to make the most of Gaudi2 is very easy with [Optimum Habana](https://huggingface.co/docs/optimum/habana/index), which enables users to port Transformers-compatible scripts to Gaudi with just a 2-line change.

## Benchmark

> To benchmark training, we are going to fine-tune BridgeTower Large using [this checkpoint](https://huggingface.co/BridgeTower/bridgetower-large-itm-mlm-itc), which contains 955M parameters with all its pretraining heads. It was pretrained on English language using masked language modeling, image-text matching and image-text contrastive loss on [Conceptual Captions](https://huggingface.co/datasets/conceptual_captions), [SBU Captions](https://huggingface.co/datasets/sbu_captions), [MSCOCO Captions](https://huggingface.co/datasets/HuggingFaceM4/COCO) and [Visual Genome](https://huggingface.co/datasets/visual_genome).

Reviewer: ```suggestion
To benchmark training, we are going to fine-tune a [BridgeTower Large checkpoint](https://huggingface.co/BridgeTower/bridgetower-large-itm-mlm-itc) consisting of 955M parameters. This checkpoint was pretrained on English language using masked language modeling, image-text matching and image-text contrastive loss on [Conceptual Captions](https://huggingface.co/datasets/conceptual_captions), [SBU Captions](https://huggingface.co/datasets/sbu_captions), [MSCOCO Captions](https://huggingface.co/datasets/HuggingFaceM4/COCO) and [Visual Genome](https://huggingface.co/datasets/visual_genome).
```

Removed the pretraining heads comment, not sure it's relevant.

> This checkpoint is fine-tuned on the [New Yorker Caption Contest dataset](https://huggingface.co/datasets/jmhessel/newyorker_caption_contest) which consists of cartoons from The New Yorker and the most voted captions.

Reviewer: ```suggestion
We will further fine-tune this checkpoint on the [New Yorker Caption Contest dataset](https://huggingface.co/datasets/jmhessel/newyorker_caption_contest) which consists of cartoons from The New Yorker and the most voted captions.
```

Hyperparameters are all the same for both accelerators, except the batch size: we managed to fit 40 samples on Gaudi2 against 32 on A100. You can check them out [here](https://huggingface.co/regisss/bridgetower-newyorker-gaudi2-8x#training-hyperparameters) for Gaudi2 and [there](https://huggingface.co/regisss/bridgetower-newyorker-a100-8x#training-hyperparameters) for A100.

> **When dealing with datasets involving images, data loading is a usual bottleneck** because many costly operations are computed on CPU (image decoding, image augmentations) and then full images are sent to devices. Ideally, *we would like to send only raw bytes to devices and then perform decoding and various image transformations on device*. But let's see first how to *easily* allocate more resources to data loading for accelerating your runs.

Reviewer: ```suggestion
**When dealing with datasets involving images, data loading is frequently a bottleneck** because many costly operations are computed on CPU (image decoding, image augmentations) and then full images are sent to the training devices. Ideally, *we would like to send only raw bytes to devices and then perform decoding and various image transformations on device*. But let's see first how to *easily* allocate more resources to data loading for accelerating your runs.
```

### Making use of `dataloader_num_workers`

As explained above, data loading is a compute-intensive task when dealing with images. When it is done on CPU, a quick way to speed it up would be to allocate more subprocesses for data loading. This is very easy to do as Transformers' TrainingArguments (and thus Optimum Habana's GaudiTrainingArguments) enables us to easily do it with the `dataloader_num_workers=N` argument with *N* the number of subprocesses allocated on CPU for data loading.

The default is 0, which means that data is loaded in the main process. This may not be optimal as the main process has many things to manage. We can set it to 1 to have one fully dedicated subprocess for data loading. When several subprocesses are allocated, each one of them will be responsible for preparing a batch. This means that RAM consumption will increase with the number of workers. One recommendation would be to set it to the number of CPU cores, but those cores may not be fully free so you will have to try it out to find the best configuration.

Let's run the two following experiments:

> - a mixed-precision (*bfloat16*/*float*) run distributed on 8 devices where data loading is performed by the same process as everything else (i.e. `dataloader_num_workers=0`)

Reviewer: ```suggestion
- a mixed-precision (*bfloat16*/*float*) run distributed across 8 devices where data loading is performed by the same process as everything else (i.e. `dataloader_num_workers=0`)
```

> - a mixed-precision (*bfloat16*/*float*) run distributed on 8 devices with 1 dedicated subprocess for data loading (i.e. `dataloader_num_workers=1`)

Reviewer: ```suggestion
- a mixed-precision (*bfloat16*/*float*) run distributed across 8 devices with 1 dedicated subprocess for data loading (i.e. `dataloader_num_workers=1`)
```

Here are the throughputs we got on Gaudi2 and A100:

| Device     | `dataloader_num_workers=0` | `dataloader_num_workers=1` |
|:----------:|:--------------------------:|:--------------------------:|
| Gaudi2 HPU | 532.4 samples/s            | 639.7 samples/s            |

> | A100 GPU   | 188.6 samples/s            | 254.7 samples/s            |

Reviewer: Doesn't this depend on the CPU characteristics of the host environment? We should maybe detail the type of CPU that was used for the A100 test.

Author: True. Actually, it depends on the CPU even in the single case `dataloader_num_workers=0`. Let's see if I manage to get access to a A100 instance on GCP to get the CPU config.

We first see that **Gaudi2 is x2.51 faster than A100** with `dataloader_num_workers=1` and x2.82 faster with `dataloader_num_workers=0`, which is even better than [the speedups we previously reported](https://huggingface.co/blog/habana-gaudi-2-benchmark)!

Second, we see that **allocating more resources for data loading can lead to easy speedups**: x1.20 on Gaudi2 and x1.35 on A100.

> We also ran experiments with several dedicated subprocesses for data loading but performance was not better than with `dataloader_num_workers=1`.

Reviewer: In both cases?

Author: Yep, making it clearer

Thus, **using `dataloader_num_workers=1` is usually a good first way of accelerating your runs involving images!**

> Tensorboard logs can be visualized [here](https://huggingface.co/regisss/bridgetower-newyorker-gaudi2-8x/tensorboard) for Gaudi2 and [there](https://huggingface.co/regisss/bridgetower-newyorker-a100-8x/tensorboard) for A100.

### Optimum Habana's fast DDP

> Before delving into how to perform hardware-accelerated data loading, let's look at another very easy way of speeding your distributed runs up on Gaudi. The new release of Optimum Habana, version 1.6.0, introduced a new feature that allows users to choose the distribution strategy to use when executing a distributed run:

Reviewer: ```suggestion
Before delving into how to perform hardware-accelerated data loading, let's look at another very easy way of speeding up your distributed runs on Gaudi. The new release of Optimum Habana, version 1.6.0, introduced a new feature that allows users to choose the distribution strategy to use:
```

> - `distribution_strategy="ddp"` to use Torch [`DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html) (DDP)

Reviewer: ```suggestion
- `distribution_strategy="ddp"` to use PyTorch [`DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html) (DDP)
```

- `distribution_strategy="fast_ddp"` to use a lighter and usually faster implementation

> Optimum Habana's fast DDP does not split parameter gradients into buckets as [DDP does](https://pytorch.org/docs/stable/notes/ddp.html#internal-design). It also uses [HPU graphs](https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_HPU_Graphs.html?highlight=hpu%20graphs) to collect gradients in all processes and then update them (after the [all_reduce](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_reduce) operation is performed) with minimal host overhead. You can check this implementation out [here](https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/distributed/fast_ddp.py).

Reviewer: ```suggestion
Optimum Habana's fast DDP does not split parameter gradients into buckets as [DDP does](https://pytorch.org/docs/stable/notes/ddp.html#internal-design). It also uses [HPU graphs](https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_HPU_Graphs.html?highlight=hpu%20graphs) to collect gradients in all processes and then update them (after the [all_reduce](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_reduce) operation is performed) with minimal host overhead. You can check this implementation [here](https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/distributed/fast_ddp.py).
```

Simply using `distribution_strategy="fast_ddp"` (and keeping `dataloader_num_workers=1`) on Gaudi2 gives us 705.9 samples/s. **This is x1.10 faster than with DDP and x2.77 faster than A100!**

> So adding just two training arguments (`dataloader_num_workers=1` and `distribution_strategy="fast_ddp"`) led to a x1.33 speedup on Gaudi2.

Reviewer: ```suggestion
So adding just two training arguments (`dataloader_num_workers=1` and `distribution_strategy="fast_ddp"`) led to a ×1.33 speedup on Gaudi2 (or ×3.7 vs the A100 single-process baseline).
```

Not sure if that's the correct number, but maybe stress the comparison with A100 again.

### Hardware-accelerated data loading with Optimum Habana

> To speed runs up even more, we are now going to move as many data loading operations as possible from the CPU to the accelerator devices. This can be done on Gaudi2 using Habana's [media pipeline](https://docs.habana.ai/en/latest/Media_Pipeline/index.html).

Reviewer: ```suggestion
For even larger speedups, we are now going to move as many data loading operations as possible from the CPU to the accelerator devices. This can be done on Gaudi2 using Habana's [media pipeline](https://docs.habana.ai/en/latest/Media_Pipeline/index.html).
```

Given a dataset, most dataloaders follow the following recipe:

1. Fetch data (e.g. where your JPEG images are stored on disk)
2. The CPU reads encoded images
3. The CPU decodes images
4. The CPU applies image transformations to augment images
5. Finally, images are sent to devices (although this is usually not done by the dataloader itself)

Instead of doing the whole process on CPU and send ready-to-train data to devices, a more efficient workflow would be to send encoded images to devices first and then perform image decoding and augmentations:

1. Same as before

> 2. Same as before

Reviewer: Not sure if it makes sense to clarify that images should be resized to the same size, still on CPU, to prepare the batch.

3. Encoded images are sent to devices
4. Devices decode images

> 5. Devices apply image transformations to augment images

Reviewer: maybe clarify "accelerator devices" somewhere around this area

That way we can benefit from the computing power of our devices to speed image decoding and transformations up.

> Note that there are two caveats to be aware of when doing this:
> - Device memory consumption will increase, so you may have to reduce your batch size if there is not enough free memory. This may mitigate the speedup brought by this approach.
> - If devices are intensively used (100% or close to it) when doing data loading on CPU, don't expect any speedup when doing it on devices as they already have their hands full.

<!-- To achieve this on Gaudi2, Habana's media pipeline enables us to:
- Initialize a media pipeline with all the operators it needs (see [here](https://docs.habana.ai/en/latest/Media_Pipeline/Operators.html#media-operators) the list of all supported operators) and define a graph so that we can specify in which order operations should be performed (e.g. reading data &rarr; decoding &rarr; cropping).
- Create a Torch dataloader with a HPU-tailored iterator. -->

> To implement this on Gaudi, we have got you covered: the [contrastive image-text example](https://github.com/huggingface/optimum-habana/tree/main/examples/contrastive-image-text) in Optimum Habana now provides a ready-to-use media pipeline that you can use with COCO-like datasets that contain text and images! You will just have to add `--mediapipe_dataloader` to your command to use it.

Reviewer: Should it be `Gaudi2` in the first line?

pip install -r requirements.txt
```

The base command line to run the script is:
```bash
python ../gaudi_spawn.py --use_mpi --world_size 8 run_bridgetower.py \
--output_dir /tmp/bridgetower-test \
--model_name_or_path BridgeTower/bridgetower-large-itm-mlm-itc \
--dataset_name jmhessel/newyorker_caption_contest --dataset_config_name matching \
--image_column image --caption_column image_description \
--remove_unused_columns=False \
--do_train --do_eval --do_predict \
--per_device_train_batch_size="40" --per_device_eval_batch_size="16" \
--num_train_epochs 5 \
--learning_rate="1e-5" \
--push_to_hub --report_to tensorboard --hub_model_id bridgetower\
--overwrite_output_dir \
--use_habana --use_lazy_mode --use_hpu_graphs_for_inference --gaudi_config_name Habana/clip \
--throughput_warmup_steps 3 \
--logging_steps 10
```
which corresponds to the case `--dataloader_num_workers 0`. You can then add `--dataloader_num_workers 1`, `--distribution_strategy fast_ddp` and `--mediapipe_dataloader` to test other configurations.

> To push your model and Tensorboard logs to the Hugging Face Hub, you will have to log in to your account beforehand with:
> ```bash
> huggingface-cli login
> ```

> For A100, you can use the same `run_bridgetower.py` script with a couple of small changes:

Author: Do you mean something like "A100 80GB with 8 GPUs" or rather an instance type in AWS/GCP terms (e.g. p4 for AWS)?

> For A100, you can use the same `run_bridgetower.py` script with a couple of small changes:

Reviewer: Yes, I was referring to the instance name so people can replicate the benchmark if they want.
